text/0000-named-blocks.md

```diff
@@ -0,0 +1,635 @@
+
+```html
+{{#x-modal}}
+  <@header as |title|>
+    Header {{title}}
+  </@header>
+
+  <@body>
+    Body
+  </@body>
+
+  <@footer>
+    Footer
+  </@footer>
+{{/x-modal}}
+```
+
+Given the above invocation, here's how you could render these blocks:
+
+```
+{{! within ember-component layout }}
+{{header "ima title"}}
+{{body}}
+{{footer}}
+
+{{! within glimmer-component layout }}
+{{@header "ima title"}}
+{{@body}}
+{{@footer}}
+```
+
+Both of these Ember/Glimmer layouts would render:
+
+```
+Header ima title
+Body
+Footer
+```
+
+The mental modal here is that is that for ECs, named blocks are
+set/bound as a properties on the instance, which we're rendering the
+same way we always rendering properties on the instance. For GCs, blocks
+are just args that we're rendering with the standard @arg syntax
+
+#### Unified Renderable Syntax: primitives
+
+One nice thing about this syntax is that it works with any kind of
+"renderable", including simple strings, such that the following
+invocation for `x-modal` would also be supported:
+
+```html
+{{#x-modal footer="ima footer"}}
+  <@header as |title|>
+    Header {{title}}
+  </@header>
+
+  <@main>
+    Main
+  </@main>
+{{/x-modal}}
+```
+
+In the above invocation, we pass `footer` in as a string rather than a
+block, which still renders just fine with the Unified Renderable syntax.
+
+This helps to unify the syntax for rendering dynamic values to DOM
+and supports a common workflow where string args can be promoted
+to full-on blocks without having to rework the component code
+to support an alternative/parallel API.
+
+#### Unified Renderable Syntax: component factories
+
+The following invocation using component factories is also supported:
+
+```html
+{{#x-modal
+     header=(component 'my-modal-header')
+     footer=(component 'my-modal-footer')}}
+  <@main>
+    Main
+  </@main>
+{{/x-modal}}
+```
+
+This demonstrates that the unified renderable syntax is also capable of
+rendering component factories (previously only renderable via
+`{{component header}}`).
+
+Note since we're passing a positional param `"ima title"` to `header`,
+the `my-modal-header` component would only be able to access that param if it were
+using the `positionalParams` API with (`reopenClass`), which is a bit of
+a clunky / pro-user API.
+
+As a component author, if you want to write your components to support
+passing data to both blocks (which accept positional params) and
+components (which accept KV pairs), you can pass in both formats
+of the same data, e.g.:
+
+```
+// Ember.component layout
+{{header headerTitle title=headerTitle}}
+
+// Glimmer Component layout
+{{@header headerTitle title=headerTitle}}
+```
+
+#### OPEN QUESTION: "calling" string args
+
+How should we handle `{{@header some blockparam data}}`
+when `@header` is a non-"callable", simple value like a string? Should
+we loudly error? Should it just ignore those "extra" args and render
```

Comment:
I don't think this can be compared to the component case, which fails for the same reason injecting an unknown service should fail. This is IMO more like how `{{yield}}` no-ops when `hasBlock` is false; we don't require you to wrap all your yields in `{{#if hasBlock}}` because, well, that would be miserable.